op-e2e/actions/action.go

```diff
@@ -0,0 +1,62 @@
+package actions
+
+import (
+	"context"
+
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils"
+)
+
+// Testing is an interface to Go-like testing,
+// extended with a context getter for the test runner to shut down individual actions without interrupting the test,
+// and a signaling function for when an invalid action is hit.
+// This helps custom test runners navigate slow or invalid actions, e.g. during fuzzing.
+type Testing interface {
+	e2eutils.TestingBase
+	// Ctx shares a context to execute an action with, the test runner may interrupt the action without stopping the test.
+	Ctx() context.Context
+	// InvalidAction indicates the failure is due to action incompatibility, does not stop the test.
+	InvalidAction(format string, args ...any)
+}
+
+// Action is a function that may change the state of one or more actors or check their state.
+// Action definitions are meant to be very small building blocks,
+// and then composed into larger patterns to write more elaborate tests.
+type Action func(t Testing)
+
+// ActionStatus defines the state of an action, to make a basic distinction between InvalidAction() and other calls.
+type ActionStatus uint
+
+const (
+	ActionOK ActionStatus = iota
+	ActionInvalid
+)
+
+// defaultTesting is a simple implementation of Testing that takes standard Go testing framework,
+// and handles invalid actions as errors, and exposes a Reset function to change the context and action state,
+// to recover after an invalid action or cancelled context.
+type defaultTesting struct {
+	e2eutils.TestingBase
+	ctx   context.Context
+	state ActionStatus
+}
+
+func (st *defaultTesting) Ctx() context.Context {
+	return st.ctx
+}
+
+func (st *defaultTesting) InvalidAction(format string, args ...any) {
```

Comment:
Needs a docstring on usage (`format` + args)

Reply:
Added more docstrings

Reply:
Specifically here it'd be helpful to describe that format gets passed to Golang's fmt and must match the provided args. When directly working with fmt.Printf/Errorf there's good tooling to make sure that the args match the format string, but I'm not sure what will happen when there's a layer of indirection.

Reply:
It behaves the same, if it's wrong it will also behave the same. Tooling might not pick up on it, but I don't see a better way of doing this while also tracking invalid calls for fuzzing purposes.